### DISCUSSION TITLE
Are Spice clients / channels designed to be long-lived?

### DISCUSSION THREAD

**Discussion by theronic** [authzed/spicedb#1900] [Q&A] [answered]
What is the designed life-time of a client?
Spice moans at me when I don't release a channel. Should I be creating & shutting down the channel for every request? Seems like there would be overhead, or is there some pooling?

**Comment by vroldanbet** [comment, marked as answer]
This is all basically details of how to handle gRPC connections: https://grpc.io/docs/what-is-grpc/core-concepts/#channels

>A gRPC channel provides a connection to a gRPC server on a specified host and port. It is used when creating a client stub. Clients can specify channel arguments to modify gRPC’s default behavior, such as switching message compression on or off. A channel has state, including connected and idle.
>How gRPC deals with closing a channel is language-dependent. Some languages also permit querying channel state.

A channel is created when you create a client, which should be long-lived. It may be language-dependent but those are typically thread-safe. The recommendation is to create one client to handle all the SpiceDB communication for your application, once you hit the limits of a single channel. You can see more in the [gRPC best practices](https://grpc.io/docs/guides/performance/#general)

No need for pooling because the underlying channel does H2, which in itself does multiplexing. You may eventually hit some limits in the channel. The best practices page says:

>Each gRPC channel uses 0 or more HTTP/2 connections and each connection usually has a limit on the number of concurrent streams. When the number of active RPCs on the connection reaches this limit, additional RPCs are queued in the client and must wait for active RPCs to finish before they are sent. Applications with high load or long-lived streaming RPCs might see performance issues because of this queueing. There are two possible solutions:

>Create a separate channel for each area of high load in the application.

>Use a pool of gRPC channels to distribute RPCs over multiple connections (channels must have different channel args to prevent re-use so define a use-specific channel arg such as channel number).
